amaranth/lib/fixed.py

```diff
@@ -0,0 +1,328 @@
+
+    def __call__(self, target):
+        return Value(self, target)
+
+    def min(self):
+        c = Const(0, self)
+        c._value = c._min_value()
+        return c
+
+    def max(self):
+        c = Const(0, self)
+        c._value = c._max_value()
+        return c
+
+    def from_bits(self, raw):
+        c = Const(0, self)
+        c._value = raw
+        if self.signed and raw > c._max_value():
+            # 2s complement signed value, but `raw` was unsigned.
+            c._value = c._min_value() + c._value - c._max_value() - 1
+        if c._value < c._min_value() or c._value > c._max_value():
+            raise ValueError(
+                f"{raw} outside expected range {c._min_value()}, {c._max_value()}")
+        return c
+
+    def __repr__(self):
+        return f"fixed.Shape({self._storage_shape}, f_bits={self.f_bits})"
+
+
+class SQ(Shape):
+    def __init__(self, i_bits, f_bits):
+        super().__init__(hdl.Shape(i_bits + f_bits, signed=True), f_bits)
+
+
+class UQ(Shape):
+    def __init__(self, i_bits, f_bits):
+        super().__init__(hdl.Shape(i_bits + f_bits, signed=False), f_bits)
+
+
+class Value(hdl.ValueCastable):
+    def __init__(self, shape, target):
+        self._shape = shape
+        if self.signed and not target.shape().signed:
+            # When methods bit-pick or concatenate to
+            # the _target of a Value, and then use this
+            # to reconstruct a Value, we may lose the
+            # signedness of its underlying _target.
+            self._target = target.as_signed()
+        else:
+            self._target = target
+
+    @property
+    def signed(self):
+        return self._shape.signed
+
+    @staticmethod
+    def cast(value, f_bits=0):
+        return Shape.cast(value.shape(), f_bits)(value)
+
+    @property
+    def i_bits(self):
+        return self._shape.i_bits
+
+    @property
+    def f_bits(self):
+        return self._shape.f_bits
+
+    def shape(self):
+        return self._shape
+
+    def as_value(self):
+        return self._target
+
+    def eq(self, other):
+        if isinstance(other, hdl.Value):
+            return self.as_value().eq(other)
+        elif isinstance(other, int) or isinstance(other, float):
+            other = Const(other, self.shape())
+        elif not isinstance(other, Value):
+            raise TypeError(f"Object {other!r} cannot be converted to a fixed.Value")
+        other = other.reshape(self.f_bits)
+        return self.as_value().eq(other.as_value())
+
+    def reshape(self, f_bits):
+        # If we're increasing precision, extend with more fractional bits. If we're
+        # reducing precision, truncate bits.
+        shape = hdl.Shape(self.i_bits + f_bits, signed=self.signed)
+        if f_bits > self.f_bits:
+            result = Shape(shape, f_bits)(hdl.Cat(hdl.Const(0, f_bits - self.f_bits), self.as_value()))
+        else:
+            result = Shape(shape, f_bits)(self.as_value()[self.f_bits - f_bits:])
```

Comment:
```suggestion
        elif f_bits < self.f_bits:
            result = Shape(shape, f_bits)(self.as_value()[self.f_bits - f_bits:])
        else:
            result = Shape(shape, f_bits)(self.as_value())
```
This suggestion avoids unnecessary slicing, which simplifies the representation:
- before: `fixed.SQ(3, 3) (+ (s (slice (const 5'sd-9) 0:5)) (cat (const 3'd0) (const 1'd1)))`
- after: `fixed.SQ(3, 3) (+ (const 5'sd-9) (cat (const 3'd0) (const 1'd1)))`

Reply:
agree, thanks!